bip32: bitcoin address derivation from extended keys

yosoyubik opened the pull request:
Fixes #2877.

It updates +decompress-point:secp256k1 to use elliptic curve cryptography over the finite field p, using modulo arithmetic.
The fact that p "[...] satisfy the equation, p mod 4 === 3", allows us to "skip the somewhat complex general purpose Tonelli-Shanks algorithm, and use a simple identity to find the square roots." [1]

It also allows key derivation for mainnet, testnet and regtest, both for extended keys and addresses.

@Fang- 

- [1] https://stackoverflow.com/a/30431547

belisarius222 commented on pkg/arvo/sys/zuse.hoon:
```diff
@@ -4386,12 +4386,15 @@
+        ::
+        =?  y  !=((sub s 2) (mod y 2))
+          (sub ^p y)
         :-  x
```
These lines can be cleaned up a bit:
```hoon
~|  [`@ux`s `@ux`dat]
?>  |(=(2 s) =(3 s))
[x y]
```
It might be better to assert about `s` just after assigning it, though.

yosoyubik commented:
@belisarius222 I might be reading this wrong since I've never worked with jets, but,  although [+secp](https://github.com/urbit/urbit/pull/1019/commits/fd78579e24cd8fe697fac347d0c0c30efc503361) is jetted, only [three arms](https://github.com/urbit/urbit/blob/790ae3dadc6008a03541537f85021bb8af14e1be/pkg/urbit/jets/tree.c#L329) from the `+helper` core have been registered:  `+make-k`, `+ecdsa-raw-sign`, `+ecdsa-raw-recover`. 

These seem to be missing: 
```
+compress-point
+serialize-point
+decompress-point
+priv-to-pub
```

belisarius222 commented:
Hmm, any idea whether that's enough to guarantee a constant-time algorithm?  We usually like to have top-level jets for crypto so we don't accidentally introduce timing attack vulnerabilities in Nock code.

—
~rovnys-ricfer
https://urbit.org

On Tue, Jun 09, 2020 at 11:37 PM, jose < notifications@github.com > wrote:

> 
> 
> 
> @ belisarius222 ( https://github.com/belisarius222 ) I might be reading
> this wrong since I've never worked with jets, but, although +secp (
> https://github.com/urbit/urbit/pull/1019/commits/fd78579e24cd8fe697fac347d0c0c30efc503361
> ) is jetted, only three arms (
> https://github.com/urbit/urbit/blob/790ae3dadc6008a03541537f85021bb8af14e1be/pkg/urbit/jets/tree.c#L329
> ) from the +helper core have been registered: +make-k , +ecdsa-raw-sign , +ecdsa-raw-recover
> .
> 
> 
> 
> These seem to be missing:
> 
> +compress-point
> +serialize-point
> +decompress-point
> +priv-to-pub
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (
> https://github.com/urbit/urbit/pull/2990#issuecomment-641700486 ) , or unsubscribe
> (
> https://github.com/notifications/unsubscribe-auth/AAGVR5NGGQQNC3423PSUNQDRV35XTANCNFSM4NXI5NRA
> ).
> 
> 
>

yosoyubik commented:
@belisarius222 I have addressed your feedback, let me know if it looks ok.

regarding timing attacks, I don't know if not having jets for these arms will cause any issues... sorry!

vvisigoth commented:
@frodwith since you're dealing with crypto jets now, I'd be interested in your take on whether we need top-level jets for these arms?

frodwith commented:
If they're used directly for real crypto (i.e. they touch secrets at runtime) and not just there as helpers for the hoon, they should probably have jets. But I'm not a crypto expert.

yosoyubik commented:
@frodwith I am wondering how the work on this PR that doesn't will with the the secp core (the updates to bip32 to deal with xpub keys and different networks) plays out with your work here: https://github.com/urbit/urbit/pull/3539

frodwith commented:
@yosoyubik The API is mostly the same. The curve parameters (n, etc) are accessed a little differently and the names of one or two functions changed, that is all. Not sure if that's exactly what you're asking.

yosoyubik commented:
@frodwith i've rebased your commits on top mine, ran `-test %/tests/lib/bip32` and got `ok=%.y`, all very smooth :)

PS: I got rid of my previous commit where I fixed `+decompress-point`

cc @jalehman

joemfb commented:
@yosoyubik, the pill was probably out of date. I've merged in #3539 into `na-release/candidate`, which is the branch we'll be breaching into. You could try your `lib/bip32` changes against that branch, and, if everything's working, rebase and retarget this PR.

joemfb commented:
This CI failure is spurious -- tarball uploads s/b disabled in PR builds.

yosoyubik commented:
Thanks @joemfb!

Fang- reviewed (approved):
This looks good to me. Thanks @yosoyubik. (:
